client.go

```diff
@@ -1,137 +1,185 @@
 )
 
-// JSON Object structs.
-
-type SearchCriteriaSchema struct {
-	DocTypes []string     `json:"doc_types"`
-	Filter   SearchFilter `json:"filter"`
+// Credentials are the identification required by the Iconik API
+//
+// The AppID is the application key id that you
+// get when generating an application key.
+//
+// The Token is the string generated alongside the
+// application key.
+type Credentials struct {
+	AppID string
+	Token string
 }
 
-type SearchFilter struct {
-	Operator string       `json:"operator"`
-	Terms    []FilterTerm `json:"terms"`
-}
+// IClient implements a Iconik client. Do not modify state concurrently.
+type IClient struct {
+	Credentials
 
-type FilterTerm struct {
-	Name  string `json:"name"`
-	Value string `json:"value"`
-}
+	// If true, don't retry requests if authorization has expired
+	NoRetry bool
 
-type SearchResponse struct {
-	Objects []IconikObject `json:"objects"`
-}
+	// If true, display debugging information about API calls
+	Debug bool
 
-type IconikObject struct {
-	Files []IconikFile `json:"files"`
+	// State
+	mutex      sync.Mutex
+	host       string
+	httpClient http.Client
 }
 
-type IconikFile struct {
-	Name string `json:"name"`
+// NewIClient creates a new Client for accessing the Iconik API.
+func NewIClient(creds Credentials, host string) (*IClient, error) {
+	if host == "" {
+		host = iconikHost
+	}
+	c := &IClient{
+		Credentials: creds,
+		host:        iconikHost,
+	}
+
+	return c, nil
 }
 
-// Iconik API Client
+// Create an authorized request using the client's credentials
+func (c *IClient) authRequest(method, apiPath string, body io.Reader, header http.Header) (*http.Request, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
 
-type IconikClient struct {
-	client *http.Client
-}
+	path := c.host + apiPath
+	req, err := http.NewRequest(method, path, body)
+	req.Header = header
 
-func (i *IconikClient) sendRequest(r *http.Request) SearchResponse {
-	response, err := i.client.Do(r)
 	if err != nil {
-		fmt.Print(err.Error())
-		os.Exit(1)
+		return nil, err
 	}
 
-	responseData, err := ioutil.ReadAll(response.Body)
-	if err != nil {
-		log.Fatal(err)
+	if c.Debug {
+		log.Printf("authRequest: %s %s\n", method, req.URL)
 	}
 
-	var responseObject SearchResponse
-	json.Unmarshal(responseData, &responseObject)
-	return responseObject
+	return req, nil
 }
 
-func makeSearchHeader() http.Header {
-	header := make(http.Header)
-	header.Add("accept", "application/json")
-	header.Add("App-Id", "bb4bb090-a100-11ec-8fe5-dac11cea8e63")
-	header.Add("Content-Type", "application/json")
-	header.Add("Auth-Token", "eyJhbGciOiJIUzI1NiIsImlhdCI6MTY0Njk3ODUyMCwiZXhwIjoxOTYyNDM4NTIwfQ.eyJpZCI6ImM0MmU4ZTA4LWExMDAtMTFlYy1hYTUwLTM2ZjFkMmVmNjY1YyJ9.Am9iOHu2FOpsMME3QiEWJv_xk4sCJYaRAjJvAaATIC8")
-	header.Add("generated_signed_url", "true")
-	return header
-}
+// Dispatch an authorized API GET request
+func (c *IClient) authGet(apiPath string, body io.Reader, header http.Header) (*http.Response, error) {
```

Comment:
again, I would shy away from putting "auth" in front of things, because every call we will ever do will have auth, so you end up with every function getting auth as the prefix of the name...

Reply:
removed auth as a prefix